### PR TITLE
Unify MaxResponseHeadersLength calculations

### LINKED ISSUE #73848: SocketsHttpHandler.MaxResponseHeadersLength can cause internal 32-bit overflow with HTTP/2
### Description

Noticed here: https://github.com/grpc/grpc-dotnet/issues/1842#issuecomment-1212059692

MaxResponseHeadersLength is in kilobytes so [internally multiplies the value by 1024](https://github.com/dotnet/runtime/blob/08a11e4a8383605087429df2477dc6a79e3ed8be/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L106).

However, the value it is stored on for HTTP/2 is 32 bit: https://github.com/dotnet/runtime/blob/08a11e4a8383605087429df2477dc6a79e3ed8be/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L88

HTTP/3 is using 64 bit:
https://github.com/dotnet/runtime/blob/80640d4d28b72ddcb3e8a69e49d88f644484b25f/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs#L41

### Reproduction Steps

```cs
new SocketsHttpHandler()
  {
    MaxResponseHeadersLength = 20 * 1024 * 1024
  };
```

Then do some HTTP/2 requests.

### Expected behavior

Either works, or provides a friendly argument exception when setting MaxResponseHeadersLength that the value exceeds the maximum size.

### Actual behavior

Internal overflow of 32-bit value, unpredictable behavior, bad error message.

```
      System.Net.Http.HttpRequestException: An error occurred while sending the request.
       ---> System.IO.IOException: The request was aborted.
       ---> System.Net.Http.HttpRequestException: The HTTP response headers length exceeded the set limit of 21474836480 bytes.
         at System.Net.Http.Http2Connection.Http2Stream.OnHeader(ReadOnlySpan`1 name, ReadOnlySpan`1 value)
         at System.Net.Http.Http2Connection.Http2Stream.System.Net.Http.IHttpHeadersHandler.OnStaticIndexedHeader(Int32 index)
         at System.Net.Http.HPack.HPackDecoder.OnIndexedHeaderField(Int32 index, IHttpHeadersHandler handler)
         at System.Net.Http.HPack.HPackDecoder.Parse(ReadOnlySpan`1 data, Int32& currentIndex, IHttpHeadersHandler handler)
         at System.Net.Http.HPack.HPackDecoder.DecodeInternal(ReadOnlySpan`1 data, IHttpHeadersHandler handler)
         at System.Net.Http.Http2Connection.ProcessHeadersFrame(FrameHeader frameHeader)
         at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
         --- End of inner exception stack trace ---
         at System.Net.Http.Http2Connection.ThrowRequestAborted(Exception innerException)
         at System.Net.Http.Http2Connection.Http2Stream.CheckResponseBodyState()
         at System.Net.Http.Http2Connection.Http2Stream.TryEnsureHeaders()
         at System.Net.Http.Http2Connection.Http2Stream.ReadResponseHeadersAsync(CancellationToken cancellationToken)
         at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
         at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
         at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
         at Grpc.Shared.TelemetryHeaderHandler.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
         at Grpc.Net.Client.Balancer.Internal.BalancerHttpHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         at Grpc.Net.Client.Internal.GrpcCall`2.RunCall(HttpRequestMessage request, Nullable`1 timeout)
```

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by MihaZupan**
Fixes #73848

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs [hunk 1]
```diff
@@ -843,10 +843,10 @@ private async ValueTask ReadHeadersAsync(long headersLength, CancellationToken c
             if (headersLength > _headerBudgetRemaining)
             {
                 _stream.Abort(QuicAbortDirection.Write, (long)Http3ErrorCode.ExcessiveLoad);
-                throw new HttpRequestException(SR.Format(SR.net_http_response_headers_exceeded_length, _connection.Pool.Settings._maxResponseHeadersLength * 1024L));
+                throw new HttpRequestException(SR.Format(SR.net_http_response_headers_exceeded_length, _connection.Pool.Settings.MaxResponseHeadersByteLength));
             }
 
-            _headerBudgetRemaining -= headersLength;
+            _headerBudgetRemaining -= (int)headersLength;
```
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:849]: Is the long guaranteed to be in range? If yes, why is it a long?

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:849]: Oh, I see, the budget is an int and we would have thrown if the Headers length was longer than it.

**Review by stephentoub — APPROVED** [review]
Are we missing a test?

**Comment by MihaZupan** [comment]
We have an existing test that checks that we enforce the limit, but we don't check inputs that would overflow. Added that.

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.MaxResponseHeadersLength.cs [hunk 2]
```diff
@@ -66,124 +66,83 @@ public async Task SetAfterUse_Throws()
             server => server.AcceptConnectionSendResponseAndCloseAsync());
         }
 
-        [OuterLoop]
-        [Fact]
-        public async Task InfiniteSingleHeader_ThrowsException()
+        [Theory]
+        [InlineData(1)]
+        [InlineData(15)]
+        public async Task LargeSingleHeader_ThrowsException(int? maxResponseHeadersLength)
```
Comment (JamesNK) [inline src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.MaxResponseHeadersLength.cs:72]: All the test parameters have a value. Does this need to be nullable?

Comment (MihaZupan) [reply src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.MaxResponseHeadersLength.cs]: Not anymore (I was initially testing with the default value as well)

**Review by JamesNK — APPROVED** [review]

**Comment by stephentoub** [comment]
The tests are failing with the WinHTTP variants.

**Review by stephentoub — APPROVED** [review]

**Pull request merged**